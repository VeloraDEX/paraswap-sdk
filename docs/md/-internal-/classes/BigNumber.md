[**@paraswap/sdk**](../../README.md) • **Docs**

***

[@paraswap/sdk](../../globals.md) / [\<internal\>](../README.md) / BigNumber

# Class: BigNumber

## Implements

- [`Hexable`](../interfaces/Hexable.md)

## Constructors

### new BigNumber()

> **new BigNumber**(`constructorGuard`, `hex`): [`BigNumber`](BigNumber.md)

#### Parameters

• **constructorGuard**: `any`

• **hex**: `string`

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:7

## Properties

### \_hex

> `readonly` **\_hex**: `string`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:5

***

### \_isBigNumber

> `readonly` **\_isBigNumber**: `boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:6

## Methods

### abs()

> **abs**(): [`BigNumber`](BigNumber.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:10

***

### add()

> **add**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:11

***

### and()

> **and**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:17

***

### div()

> **div**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:13

***

### eq()

> **eq**(`other`): `boolean`

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

`boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:23

***

### fromTwos()

> **fromTwos**(`value`): [`BigNumber`](BigNumber.md)

#### Parameters

• **value**: `number`

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:8

***

### gt()

> **gt**(`other`): `boolean`

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

`boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:26

***

### gte()

> **gte**(`other`): `boolean`

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

`boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:27

***

### isNegative()

> **isNegative**(): `boolean`

#### Returns

`boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:28

***

### isZero()

> **isZero**(): `boolean`

#### Returns

`boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:29

***

### lt()

> **lt**(`other`): `boolean`

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

`boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:24

***

### lte()

> **lte**(`other`): `boolean`

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

`boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:25

***

### mask()

> **mask**(`value`): [`BigNumber`](BigNumber.md)

#### Parameters

• **value**: `number`

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:20

***

### mod()

> **mod**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:15

***

### mul()

> **mul**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:14

***

### or()

> **or**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:18

***

### pow()

> **pow**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:16

***

### shl()

> **shl**(`value`): [`BigNumber`](BigNumber.md)

#### Parameters

• **value**: `number`

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:21

***

### shr()

> **shr**(`value`): [`BigNumber`](BigNumber.md)

#### Parameters

• **value**: `number`

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:22

***

### sub()

> **sub**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:12

***

### toBigInt()

> **toBigInt**(): `bigint`

#### Returns

`bigint`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:31

***

### toHexString()

> **toHexString**(): `string`

#### Returns

`string`

#### Implementation of

[`Hexable`](../interfaces/Hexable.md).[`toHexString`](../interfaces/Hexable.md#tohexstring)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:33

***

### toJSON()

> **toJSON**(`key`?): `any`

#### Parameters

• **key?**: `string`

#### Returns

`any`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:34

***

### toNumber()

> **toNumber**(): `number`

#### Returns

`number`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:30

***

### toString()

> **toString**(): `string`

Returns a string representation of an object.

#### Returns

`string`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:32

***

### toTwos()

> **toTwos**(`value`): [`BigNumber`](BigNumber.md)

#### Parameters

• **value**: `number`

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:9

***

### xor()

> **xor**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:19

***

### from()

> `static` **from**(`value`): [`BigNumber`](BigNumber.md)

#### Parameters

• **value**: `any`

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:35

***

### isBigNumber()

> `static` **isBigNumber**(`value`): `value is BigNumber`

#### Parameters

• **value**: `any`

#### Returns

`value is BigNumber`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:36
