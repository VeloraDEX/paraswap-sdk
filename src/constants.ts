import { EnumerateLiteral, APIVersion } from './types';

import type { ContractMethod } from '@paraswap/core';

export {
  SwapSide,
  ContractMethod,
  ContractMethodV5,
  ContractMethodV6,
} from '@paraswap/core';

// allows to pass by name instead of importing enum first
export type ContractMethodByName = EnumerateLiteral<typeof ContractMethod>;

export const API_URL = 'https://api.paraswap.io';
export const DEFAULT_VERSION = '6.2' satisfies APIVersion;
