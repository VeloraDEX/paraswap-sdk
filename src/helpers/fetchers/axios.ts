import type { ExtraFetchParams, FetcherFunction } from '../../types';
import type { AxiosStatic } from 'axios';
import { FetcherError } from '../misc';

export type AxiosRequirement = Pick<AxiosStatic, 'request' | 'isAxiosError'>;

export const constructFetcher =
  (axios: AxiosRequirement, extra?: ExtraFetchParams): FetcherFunction =>
  async (params) => {
    try {
      const { requestParams, ...rest } = params;
      // adding apiKey to headers if it's provided
      const headers = extra?.apiKey
        ? {
            'X-API-KEY': extra.apiKey,
            ...rest.headers,
            ...requestParams?.headers,
          }
        : { ...rest.headers, ...requestParams };

      const allParams = { ...rest, ...requestParams, headers };

      const { data } = await axios.request(allParams);

      return data;
    } catch (error: any) {
      if (axios.isAxiosError(error)) {
        // attach a more meaningful message and other props
        throw new FetcherError(error);
      }

      throw error;
    }
  };
