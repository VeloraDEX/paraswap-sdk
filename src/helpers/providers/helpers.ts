import { assert } from 'ts-essentials';
import { TypedDataField } from '../../methods/common/orders/buildOrderData';

// regex from @ethersproject/hash TypedDataEncoder.constructor
// may be overly strict, but reliable
const baseTypeRegex = /^([^\x5b]*)(\x5b|$)/;

export function findPrimaryType(
  types: Record<string, TypedDataField[]>
): string {
  const candidates = Object.keys(types);
  const candidatesSet = new Set(candidates);

  candidates.forEach((candidate) => {
    const typedDataFields = types[candidate];
    if (!typedDataFields) return;

    typedDataFields.forEach(({ type }) => {
      // Get the base type (drop any array specifiers)
      const baseType = type.match(baseTypeRegex)?.[1];
      if (!baseType) return;

      // if type was referred to as a child of another type, it can't be the primaryType
      candidatesSet.delete(baseType);
    });
  });

  const [primaryType] = Array.from(candidatesSet);

  assert(
    primaryType,
    `No primary type found in SignableTypedData types, ${JSON.stringify(types)}`
  );

  return primaryType;
}
