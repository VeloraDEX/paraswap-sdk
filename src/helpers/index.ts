import { EthersV5ProviderDeps } from './providers/ethers';
import { EthersV6ProviderDeps } from './providers/ethersV6';
export { constructFetcher as constructAxiosFetcher } from './fetchers/axios';
export { constructFetcher as constructFetchFetcher } from './fetchers/fetch';
export {
  constructEthersV5ContractCaller as constructEthersContractCaller,
  constructEthersV5ContractCaller,
  EthersV5ProviderDeps,
} from './providers/ethers';
export {
  constructContractCaller as constructEthersV6ContractCaller,
  EthersV6ProviderDeps,
} from './providers/ethersV6';
export type EthersProviderDeps = EthersV5ProviderDeps | EthersV6ProviderDeps;

export {
  constructContractCaller as constructWeb3ContractCaller,
  Web3UnpromiEvent,
} from './providers/web3';
export {
  constructContractCaller as constructViemContractCaller,
  MinViemClient,
  txParamsToViemTxParams,
} from './providers/viem';
export { constructToken } from './token';
export { isFetcherError } from './misc';
export type { FetcherError, FetcherErrorInterface } from './misc';
