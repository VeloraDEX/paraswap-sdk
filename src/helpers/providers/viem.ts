import type {
  Address,
  ContractCallerFunctions,
  ContractCallStaticInput,
  SignTypedDataContractCallerFn,
  StaticContractCallerFn,
  TransactionContractCallerFn,
} from '../../types';
import { assert, MarkOptional } from 'ts-essentials';

import type {
  PublicActions,
  WalletActions,
  Hex,
  BlockTag,
  Chain,
  Transport,
  TypedDataDomain,
  Account,
} from 'viem';
import { findPrimaryType } from './helpers';
import { TransactionParams } from '../../methods/swap/transaction';

export type MinViemClient<
  account extends Account | undefined = Account | undefined
> = Partial<
  Pick<PublicActions<Transport, Chain>, 'readContract'> &
    Pick<WalletActions<Chain, account>, 'writeContract' | 'signTypedData'>
> & { account?: account };

export const constructContractCaller = (
  viemClient: MinViemClient,
  account?: Address
): ContractCallerFunctions<Hex> => {
  const staticCall: StaticContractCallerFn = async <
    T,
    M extends string = string
  >(
    params: ContractCallStaticInput<M>
  ): Promise<T> => {
    assert(
      viemClient.readContract,
      'Viem client must have readContract Public Action'
    );
    const { address, abi, contractMethod, args, overrides } = params;

    const { block } = overrides;

    const { blockNumber, blockTag } =
      typeof block === 'number'
        ? { blockNumber: BigInt(block) }
        : { blockTag: block as BlockTag };

    const result = viemClient.readContract({
      address: address as Hex,
      abi,
      functionName: contractMethod,
      args,
      blockTag,
      blockNumber,
    });

    return result as T; // can't get specific type for generic abi & method
  };

  const transactCall: TransactionContractCallerFn<Hex> = async (params) => {
    assert(account, 'account must be specified to create a signer');
    assert(
      isViemClientWithAccount(viemClient),
      'account must be specified to create a signer'
    );
    assert(
      viemClient.writeContract,
      'Viem client must have writeContract Wallet Action'
    );

    const { address, abi, contractMethod, args, overrides } = params;
    console.log(
      'ðŸš€ ~ consttransactCall:TransactionContractCallerFn<Hex>= ~ overrides:',
      overrides
    );

    const gas = overrides.gas !== undefined ? BigInt(overrides.gas) : undefined;
    const gasPrice =
      overrides.gasPrice !== undefined ? BigInt(overrides.gasPrice) : undefined;
    const maxFeePerGas =
      overrides.maxFeePerGas !== undefined
        ? BigInt(overrides.maxFeePerGas)
        : undefined;
    const maxPriorityFeePerGas =
      overrides.maxPriorityFeePerGas !== undefined
        ? BigInt(overrides.maxPriorityFeePerGas)
        : undefined;

    const value =
      overrides.value !== undefined ? BigInt(overrides.value) : undefined;

    const viemTxParams = txParamsToViemTxParams({
      ...overrides,
      to: address,
      gas: overrides.gas?.toString(10),
      value: overrides.value?.toString(10),
    });

    /*     console.log('View.writeContract', {
      address: address as Hex,
      // abi,
      functionName: contractMethod,
      args: argsToViemArgs(args),
      ...viemTxParams,
      nonce: overrides.nonce,
    }); */

    /*     const t = await viemClient.writeContract({
      address: '0x6B175474E89094C44Da98b954EedeAC495271d0F',
      abi: [
        {
          constant: false,
          inputs: [
            { name: '_spender', type: 'address' },
            { name: '_value', type: 'uint256' },
          ],
          name: 'approve',
          outputs: [{ name: '', type: 'bool' }],
          payable: false,
          stateMutability: 'nonpayable',
          type: 'function',
        },
      ],
      functionName: 'approve',
      args: ['0x216b4b4ba9f3e719726886d34a177484278bfcae', 12345n],
    });

    return t; */

    const gasPriceParams =
      'maxFeePerGas' in viemTxParams
        ? {
            maxFeePerGas: viemTxParams.maxFeePerGas,
            maxPriorityFeePerGas: viemTxParams.maxPriorityFeePerGas,
          }
        : 'gasPrice' in viemTxParams
        ? { gasPrice: viemTxParams.gasPrice }
        : {};

    const txHash = await viemClient.writeContract({
      address: address as Hex,
      abi,
      functionName: contractMethod,
      args: argsToViemArgs(args),
      value: viemTxParams.value,
      gas: viemTxParams.gas,
      // ...viemTxParams,
      // account: (overrides.from || account) as Hex,
      nonce: overrides.nonce,
      ...gasPriceParams,
    });

    return txHash;
  };

  const signTypedDataCall: SignTypedDataContractCallerFn = async (
    typedData
  ) => {
    assert(account, 'account must be specified to create a signer');
    assert(
      viemClient.signTypedData,
      'Viem client must have signTypedData Wallet Action'
    );

    const { data, domain, types } = typedData;

    const primaryType = findPrimaryType(types);

    const viemDomain: TypedDataDomain = {
      chainId: domain.chainId,
      name: domain.name,
      version: domain.version,
      verifyingContract: domain.verifyingContract as Hex,
      salt: domain.salt as Hex,
    };

    const signature = await viemClient.signTypedData({
      account: account as Hex,
      domain: viemDomain,
      types,
      primaryType,
      message: data,
    });

    return signature;
  };

  return { staticCall, transactCall, signTypedDataCall };
};

type ViemTxParams = {
  to: Hex;
  data?: Hex;
  value?: bigint;
  nonce?: number;
  gas?: bigint;
} & (
  | { maxFeePerGas?: bigint; maxPriorityFeePerGas?: bigint }
  | { gasPrice?: bigint }
);

export function txParamsToViemTxParams(
  txParams: Omit<
    MarkOptional<TransactionParams, 'data' | 'value'>,
    'chainId' | 'from'
  >
): ViemTxParams {
  const gas = txParams.gas !== undefined ? BigInt(txParams.gas) : undefined;
  const gasPrice =
    txParams.gasPrice !== undefined ? BigInt(txParams.gasPrice) : undefined;
  const maxFeePerGas =
    txParams.maxFeePerGas !== undefined
      ? BigInt(txParams.maxFeePerGas)
      : undefined;
  const maxPriorityFeePerGas =
    txParams.maxPriorityFeePerGas !== undefined
      ? BigInt(txParams.maxPriorityFeePerGas)
      : undefined;

  const viemTxParams = {
    to: txParams.to as Hex,
    data: txParams.data as Hex,
    value: txParams.value !== undefined ? BigInt(txParams.value) : undefined,
    gas,
    ...(txParams.maxFeePerGas !== undefined
      ? { maxFeePerGas, maxPriorityFeePerGas }
      : { gasPrice }),
  };
  return viemTxParams;
}

function argsToViemArgs(args: any[]): any[] {
  return args.map((arg) => {
    if (typeof arg === 'string') {
      if (arg.startsWith('0x')) {
        return arg;
      }
      const asNumber = Number(arg);
      if (Number.isNaN(asNumber)) {
        return arg;
      }
      return BigInt(asNumber);
    }
    return arg;
  });
}

function isViemClientWithAccount(
  viemClient: MinViemClient<Account | undefined>
): viemClient is MinViemClient<Account> {
  return !!viemClient.account;
}
