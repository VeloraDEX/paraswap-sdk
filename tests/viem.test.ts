import * as dotenv from 'dotenv';
import { ethers } from 'ethersV5';
import axios from 'axios';
import {
  createTestClient,
  custom,
  Hex,
  publicActions,
  parseEther,
  createWalletClient,
} from 'viem';
import { privateKeyToAccount } from 'viem/accounts';
import { hardhat } from 'viem/chains';

import { constructSimpleSDK, SimpleSDK } from '../src/sdk/simple';
import BigNumber from 'bignumber.js';
import { txParamsToViemTxParams } from '../src/helpers';
import { HardhatProvider } from './helpers/hardhat';

dotenv.config();

jest.setTimeout(30 * 1000);

const chainId = 1;

const TEST_MNEMONIC =
  'radar blur cabbage chef fix engine embark joy scheme fiction master release';
const wallet = ethers.Wallet.fromMnemonic(TEST_MNEMONIC);

const ethersProvider = new ethers.providers.Web3Provider(
  HardhatProvider as any
);

const signer = wallet.connect(ethersProvider);
const senderAddress = signer.address as Hex;

const viemTestClient = createTestClient({
  chain: { ...hardhat, id: chainId }, // may need to override chainId
  mode: 'hardhat',
  transport: custom(HardhatProvider),
}).extend(publicActions);

const viemWalletClient = createWalletClient({
  // either walletClient needs to have account set at creation
  // or provider must own the account (for testing can `await viemTestClient.impersonateAccount({ address: senderAddress });`)
  // to be able to sign transactions
  account: privateKeyToAccount(wallet.privateKey as Hex),
  chain: { ...hardhat, id: chainId },
  transport: custom(HardhatProvider),
});

describe('ParaSwap SDK: contract calling methods', () => {
  let SDKwithEthers: SimpleSDK;
  let SDKwithViem: SimpleSDK;

  const ETH = '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee';
  const DAI = '0x6B175474E89094C44Da98b954EedeAC495271d0F';
  const HEX = '0x2b591e99afe9f32eaa6214f7b7629768c40eeb39';

  const srcToken = DAI;
  const destToken = HEX;

  //                                        UTC format
  const orderExpiry = Math.floor(new Date('2025-12-20').getTime() / 1000);

  const orderInput = {
    nonce: 1,
    expiry: orderExpiry,
    makerAsset: srcToken,
    takerAsset: destToken,
    makerAmount: (1e18).toString(10),
    takerAmount: (8e18).toString(10),
    maker: senderAddress,
  };

  let spender: Hex;

  beforeAll(async () => {
    await viemTestClient.setBalance({
      address: senderAddress,
      value: parseEther('10'),
    });

    SDKwithEthers = constructSimpleSDK(
      {
        chainId,
        axios,
        version: '6.2',
      },
      {
        ethersProviderOrSigner: signer,
        EthersContract: ethers.Contract,
        account: senderAddress,
      }
    );

    SDKwithViem = constructSimpleSDK(
      {
        chainId,
        axios,
        version: '6.2',
      },
      {
        viemClient: viemWalletClient,
        account: senderAddress,
      }
    );

    spender = (await SDKwithEthers.swap.getSpender()) as Hex;
  });

  async function getDaiAllowance() {
    const allowance = await viemTestClient.readContract({
      address: DAI,
      abi: [
        {
          constant: true,
          inputs: [
            { name: '_owner', type: 'address' },
            { name: '_spender', type: 'address' },
          ],
          name: 'allowance',
          outputs: [{ name: '', type: 'uint256' }],
          payable: false,
          stateMutability: 'view',
          type: 'function',
        },
      ],
      functionName: 'allowance',
      args: [senderAddress, spender],
    });

    return allowance;
  }

  async function getDaiBalance() {
    const balance = await viemTestClient.readContract({
      address: DAI,
      abi: [
        {
          constant: true,
          inputs: [{ name: '_owner', type: 'address' }],
          name: 'balanceOf',
          outputs: [{ name: 'balance', type: 'uint256' }],
          payable: false,
          stateMutability: 'view',
          type: 'function',
        },
      ],
      functionName: 'balanceOf',
      args: [senderAddress],
    });

    return balance;
  }

  test('approval with viem', async () => {
    const allowance1 = await getDaiAllowance();

    expect(allowance1).toEqual(0n);

    const txHash = await SDKwithViem.swap.approveToken('12345', DAI);

    await viemTestClient.waitForTransactionReceipt({ hash: txHash as Hex });

    const allowance2 = await getDaiAllowance();

    expect(allowance2).toEqual(12345n);
  });

  test('approveToken', async () => {
    const ethersTxHash = await SDKwithEthers.swap.approveToken('12345', DAI);
    await viemTestClient.waitForTransactionReceipt({
      hash: ethersTxHash as Hex,
    });
    const allowance1 = await getDaiAllowance();
    expect(allowance1).toEqual(12345n);

    const viemTxHash = await SDKwithViem.swap.approveToken('1234567', DAI);
    await viemTestClient.waitForTransactionReceipt({ hash: viemTxHash });

    const allowance2 = await getDaiAllowance();
    expect(allowance2).toEqual(1234567n);
  }, 120000);

  test('signOrder', async () => {
    const signableOrderData = await SDKwithEthers.limitOrders.buildLimitOrder(
      orderInput
    );

    expect(signableOrderData).toMatchSnapshot('LimitOrder to sign');

    const ethersSignature = await SDKwithEthers.limitOrders.signLimitOrder(
      signableOrderData
    );

    const viemSignature = await SDKwithViem.limitOrders.signLimitOrder(
      signableOrderData
    );

    expect(viemSignature).toMatchInlineSnapshot(
      `"0x18b022691daab1d8a3486aab5a006f2e98932b1c2fe4f04726c766e7a4e6d4935cbbf6d03ef945f23bef5cf4e250086f14581b1b8097f6c3ffd62653c8b2454b1b"`
    );
    expect(ethersSignature).toEqual(viemSignature);
  });

  test('market swap', async () => {
    const srcAmount = (1 * 1e18).toString();
    const srcToken = ETH;
    const destToken = DAI;
    const userAddress = senderAddress;

    const priceRoute = await SDKwithViem.swap.getRate({
      srcToken,
      destToken,
      amount: srcAmount,
      userAddress,
      options: {
        includeDEXS: ['Uniswap', 'UniswapV2', 'Balancer'],
      },
    });

    const destAmount = new BigNumber(priceRoute.destAmount)
      .times(0.99) // slippage 1%
      .toFixed(0);

    const txParams = await SDKwithViem.swap.buildTx(
      {
        srcToken,
        destToken,
        srcAmount,
        destAmount,
        priceRoute,
        userAddress,
      },
      { ignoreChecks: true }
    );

    const viemTxParams = txParamsToViemTxParams(txParams);

    const balanceDAI1 = await getDaiBalance();
    expect(balanceDAI1).toEqual(0n);
    const balanceETH1 = await viemTestClient.getBalance({
      address: senderAddress,
    });

    const txHash = await viemWalletClient.sendTransaction(viemTxParams);
    await viemTestClient.waitForTransactionReceipt({
      hash: txHash,
    });

    const balanceDAI2 = await getDaiBalance();
    // destAmount + a bit more if lucky enough
    expect(balanceDAI2).toBeGreaterThanOrEqual(BigInt(destAmount));

    const balanceETH2 = await viemTestClient.getBalance({
      address: senderAddress,
    });
    // bal1 - srcAmount - some gas
    expect(balanceETH2).toBeLessThan(balanceETH1 - BigInt(srcAmount));
  });
});
