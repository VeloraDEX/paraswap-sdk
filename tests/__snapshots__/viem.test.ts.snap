// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParaSwap SDK: contract calling methods signOrder: LimitOrder to sign 1`] = `
{
  "data": {
    "expiry": 1766188800,
    "maker": "0xaC39b311DCEb2A4b2f5d8461c1cdaF756F4F7Ae9",
    "makerAmount": "1000000000000000000",
    "makerAsset": "0x6B175474E89094C44Da98b954EedeAC495271d0F",
    "nonceAndMeta": "1461501637330902918203684832716283019655932542976",
    "taker": "0x0000000000000000000000000000000000000000",
    "takerAmount": "8000000000000000000",
    "takerAsset": "0x2b591e99afe9f32eaa6214f7b7629768c40eeb39",
  },
  "domain": {
    "chainId": 1,
    "name": "AUGUSTUS RFQ",
    "verifyingContract": "0xe92b586627ccA7a83dC919cc7127196d70f55a06",
    "version": "1",
  },
  "types": {
    "Order": [
      {
        "name": "nonceAndMeta",
        "type": "uint256",
      },
      {
        "name": "expiry",
        "type": "uint128",
      },
      {
        "name": "makerAsset",
        "type": "address",
      },
      {
        "name": "takerAsset",
        "type": "address",
      },
      {
        "name": "maker",
        "type": "address",
      },
      {
        "name": "taker",
        "type": "address",
      },
      {
        "name": "makerAmount",
        "type": "uint256",
      },
      {
        "name": "takerAmount",
        "type": "uint256",
      },
    ],
  },
}
`;
